fdbserver/storageserver.actor.cpp

```diff
@@ -3739,16 +3777,29 @@ ACTOR Future<Void> replaceInterface( StorageServer* self, StorageServerInterface
 	return Void();
 }
 
-ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered )
+ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered, Reference<ClusterConnectionFile> connFile)
 {
 	state StorageServer self(persistentData, db, ssi);
 	self.folder = folder;
 	self.sk = serverKeysPrefixFor( self.thisServerID ).withPrefix(systemKeys.begin);  // FFFF/serverKeys/[this server]/
 	try {
 		state double start = now();
 		TraceEvent("StorageServerRebootStart", self.thisServerID);
-		wait(self.storage.init());
-		wait(self.storage.commit()); //after a rollback there might be uncommitted changes.
+
+		state Future<bool> dispose = memoryStoreRecover (persistentData, connFile, self.thisServerID);
+        wait(self.storage.init());
+		//after a rollback there might be uncommitted changes.
+		//for memory storage engine type, wait until recovery is done before commit
+		state Future<Void> committed = self.storage.commit();
+		wait(success(dispose) || success(committed));
```

Comment:
if dispose would return a `Future<Void>` you could instead use a choose-when construct here (and you can get rid of the two state variables):

```
choose {
    when (wait(self.storage.commit())) {}
    when (wait(memoryStoreRecover (persistentData, connFile, self.thisServerID))) {
        TraceEvent("DisposeStorageServer", self.thisServerID);
         throw worker_removed();
    }
}
```

If you do this you also don't need to explicitly cancel the `dispose` Future.

Reply:
i see. However, one of my concern is, if `type != memory` the method memoryStoreRecover() would immediately return `Never()`, but in this case, we don't want to *throw worker_removed();* right?

Reply:
Never mind, figure it out. 
push the changes into branch, thank you!